### PR TITLE
Revise gRPC error statuses and details in meta (Redis/Cassandra)

### PR DESCRIPTION
**Pull request by rinx**
Signed-off-by: Rintaro Okamura <rintaro.okamura@gmail.com>

### Description:

- Added gRPC error details in meta Redis/Cassandra.

Reference:
- https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto

### Related Issue:

Nothing

### How Has This Been Tested?:

Nothing

### Environment:

- Go Version: 1.16
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.3

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [X] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [X] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [X] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by kpango** [comment]
In the case of Unavailale or Unkonwn errors, it would be better to return the target IP of the failed connection as the resource name in errdetails.ResourceInfo and additionally errdetails.RequestInfo containing the key of the request.

**Comment by rinx** [comment]
Agree. I'll revise it.

**Comment by rinx** [comment]
Both of gocql and go-redis don't provide informations about requesting Cassandra/Redis hosts. So we cannot embed IPs of these hosts into ResourceInfo.

**Comment by rinx** [comment]
I've just noticed that some of the Cassandra errors can be categorized as the gRPC errors other than Unknown error. So I'll tackled with it.

**Comment by rinx** [comment]
Now ready to be reviewed.
/rebase

File: pkg/meta/cassandra/handler/grpc/handler.go [hunk 1]
```diff
@@ -56,24 +75,79 @@ func (s *server) GetMeta(ctx context.Context, key *payload.Meta_Key) (*payload.M
...
+					Owner:        errdetails.ValdResourceOwner,
+					Description:  err.Error(),
+				}, info.Get())
+			if span != nil {
+				span.SetStatus(trace.StatusCodeInternal(err.Error()))
+			}
+			return nil, err
+		case isFailedPrecondition(err):
+			log.Warnf("[GetMeta]\tfailed precondition\t%+v", err)
+			err = status.WrapWithInternal(fmt.Sprintf("GetMeta API: failed precondition: key %s", key.GetKey()), err,
```
Comment (kpango) [inline pkg/meta/cassandra/handler/grpc/handler.go:123]: maybe
```suggestion
			err = status.WrapWithFailedPreconditionl(fmt.Sprintf("GetMeta API: failed precondition: key %s", key.GetKey()), err,
```

Comment (kpango) [reply pkg/meta/cassandra/handler/grpc/handler.go:123]: and how about add Precondition Error details?

File: pkg/meta/cassandra/handler/grpc/handler.go [hunk 2]
```diff
@@ -94,24 +168,79 @@ func (s *server) GetMetas(ctx context.Context, keys *payload.Meta_Keys) (mv *pay
...
+					Owner:        errdetails.ValdResourceOwner,
+					Description:  err.Error(),
+				}, info.Get())
+			if span != nil {
+				span.SetStatus(trace.StatusCodeInternal(err.Error()))
+			}
+			return nil, err
+		case isFailedPrecondition(err):
+			log.Warnf("[GetMetas]\tfailed precondition\t%+v", err)
+			err = status.WrapWithInternal(fmt.Sprintf("GetMetas API: failed precondition: keys %#v", keys.GetKeys()), err,
```
Comment (kpango) [inline pkg/meta/cassandra/handler/grpc/handler.go:216]: same

File: pkg/meta/cassandra/handler/grpc/handler.go [hunk 3]
```diff
@@ -129,24 +258,79 @@ func (s *server) GetMetaInverse(ctx context.Context, val *payload.Meta_Val) (*pa
...
+					Owner:        errdetails.ValdResourceOwner,
+					Description:  err.Error(),
+				}, info.Get())
+			if span != nil {
+				span.SetStatus(trace.StatusCodeInternal(err.Error()))
+			}
+			return nil, err
+		case isFailedPrecondition(err):
+			log.Warnf("[GetMetaInverse]\tfailed precondition\t%+v", err)
+			err = status.WrapWithInternal(fmt.Sprintf("GetMetaInverse API: failed precondition: val %s", val.GetVal()), err,
```
Comment (kpango) [inline pkg/meta/cassandra/handler/grpc/handler.go:306]: same

File: pkg/meta/cassandra/handler/grpc/handler.go [hunk 4]
```diff
@@ -167,24 +351,79 @@ func (s *server) GetMetasInverse(ctx context.Context, vals *payload.Meta_Vals) (
...
+					Owner:        errdetails.ValdResourceOwner,
+					Description:  err.Error(),
+				}, info.Get())
+			if span != nil {
+				span.SetStatus(trace.StatusCodeInternal(err.Error()))
+			}
+			return nil, err
+		case isFailedPrecondition(err):
+			log.Warnf("[GetMetasInverse]\tfailed precondition\t%+v", err)
+			err = status.WrapWithInternal(fmt.Sprintf("GetMetasInverse API: failed precondition: vals %#v", vals.GetVals()), err,
```
Comment (kpango) [inline pkg/meta/cassandra/handler/grpc/handler.go:399]: same

File: pkg/meta/cassandra/handler/grpc/handler.go [hunk 5]
```diff
@@ -199,11 +438,68 @@ func (s *server) SetMeta(ctx context.Context, kv *payload.Meta_KeyVal) (_ *paylo
...
+					Owner:        errdetails.ValdResourceOwner,
+					Description:  err.Error(),
+				}, info.Get())
+			if span != nil {
+				span.SetStatus(trace.StatusCodeInternal(err.Error()))
+			}
+			return nil, err
+		case isFailedPrecondition(err):
+			log.Warnf("[SetMeta]\tfailed precondition\t%+v", err)
+			err = status.WrapWithInternal(fmt.Sprintf("SetMeta API: failed precondition: key %s val %s", kv.GetKey(), kv.GetVal()), err,
```
Comment (kpango) [inline pkg/meta/cassandra/handler/grpc/handler.go:474]: same

File: pkg/meta/cassandra/handler/grpc/handler.go [hunk 6]
```diff
@@ -221,11 +517,68 @@ func (s *server) SetMetas(ctx context.Context, kvs *payload.Meta_KeyVals) (_ *pa
...
+					Owner:        errdetails.ValdResourceOwner,
+					Description:  err.Error(),
+				}, info.Get())
+			if span != nil {
+				span.SetStatus(trace.StatusCodeInternal(err.Error()))
+			}
+			return nil, err
+		case isFailedPrecondition(err):
+			log.Warnf("[SetMetas]\tfailed precondition\t%+v", err)
+			err = status.WrapWithInternal(fmt.Sprintf("SetMetas API: failed precondition: kvs %#v", query), err,
```
Comment (kpango) [inline pkg/meta/cassandra/handler/grpc/handler.go:553]: same

File: pkg/meta/cassandra/handler/grpc/handler.go [hunk 7]
```diff
@@ -242,24 +595,79 @@ func (s *server) DeleteMeta(ctx context.Context, key *payload.Meta_Key) (*payloa
...
+					Owner:        errdetails.ValdResourceOwner,
+					Description:  err.Error(),
+				}, info.Get())
+			if span != nil {
+				span.SetStatus(trace.StatusCodeInternal(err.Error()))
+			}
+			return nil, err
+		case isFailedPrecondition(err):
+			log.Warnf("[DeleteMeta]\tfailed precondition\t%+v", err)
+			err = status.WrapWithInternal(fmt.Sprintf("DeleteMeta API: failed precondition: key %s", key.GetKey()), err,
```
Comment (kpango) [inline pkg/meta/cassandra/handler/grpc/handler.go:643]: same

File: pkg/meta/cassandra/handler/grpc/handler.go [hunk 8]
```diff
@@ -280,24 +688,79 @@ func (s *server) DeleteMetas(ctx context.Context, keys *payload.Meta_Keys) (mv *
...
+					Owner:        errdetails.ValdResourceOwner,
+					Description:  err.Error(),
+				}, info.Get())
+			if span != nil {
+				span.SetStatus(trace.StatusCodeInternal(err.Error()))
+			}
+			return nil, err
+		case isFailedPrecondition(err):
+			log.Warnf("[DeleteMetas]\tfailed precondition\t%+v", err)
+			err = status.WrapWithInternal(fmt.Sprintf("DeleteMetas API: failed precondition: keys %#v", keys.GetKeys()), err,
```
Comment (kpango) [inline pkg/meta/cassandra/handler/grpc/handler.go:736]: same

File: pkg/meta/cassandra/handler/grpc/handler.go [hunk 9]
```diff
@@ -315,24 +778,79 @@ func (s *server) DeleteMetaInverse(ctx context.Context, val *payload.Meta_Val) (
...
+					Owner:        errdetails.ValdResourceOwner,
+					Description:  err.Error(),
+				}, info.Get())
+			if span != nil {
+				span.SetStatus(trace.StatusCodeInternal(err.Error()))
+			}
+			return nil, err
+		case isFailedPrecondition(err):
+			log.Warnf("[DeleteMetaInverse]\tfailed precondition\t%+v", err)
+			err = status.WrapWithInternal(fmt.Sprintf("DeleteMetaInverse API: failed precondition: val %s", val.GetVal()), err,
```
Comment (kpango) [inline pkg/meta/cassandra/handler/grpc/handler.go:826]: same

File: pkg/meta/cassandra/handler/grpc/handler.go [hunk 10]
```diff
@@ -353,24 +871,79 @@ func (s *server) DeleteMetasInverse(ctx context.Context, vals *payload.Meta_Vals
...
+					Owner:        errdetails.ValdResourceOwner,
+					Description:  err.Error(),
+				}, info.Get())
+			if span != nil {
+				span.SetStatus(trace.StatusCodeInternal(err.Error()))
+			}
+			return nil, err
+		case isFailedPrecondition(err):
+			log.Warnf("[DeleteMetasInverse]\tfailed precondition\t%+v", err)
+			err = status.WrapWithInternal(fmt.Sprintf("DeleteMetasInverse API: failed precondition: vals %#v", vals.GetVals()), err,
```
Comment (kpango) [inline pkg/meta/cassandra/handler/grpc/handler.go:919]: same

**Review by kpango — CHANGES_REQUESTED** [review]
Others looks good to me

File: pkg/meta/cassandra/handler/grpc/handler.go (hunk 1 above)
Comment (rinx) [reply pkg/meta/cassandra/handler/grpc/handler.go:123]: oops sorry...

**Comment by rinx** [comment]
Updated.
/rebase

**Review by kpango — APPROVED** [review]
LGTM

**Pull request merged**